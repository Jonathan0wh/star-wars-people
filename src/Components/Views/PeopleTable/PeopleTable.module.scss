.column_container {
  padding: 16px;
}

.skeleton {
  width: 90vw !important;
}

@media (min-width: 992px) {
  .skeleton {
    width: 900px !important;
  }
}
