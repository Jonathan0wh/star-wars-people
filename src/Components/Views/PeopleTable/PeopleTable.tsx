import { useState } from "react";
import { useAppDispatch } from "hooks/react-redux";
import { Col, Row, Skeleton, Table, Typography } from "antd";
import { useGetPeopleQuery } from "services/swApi";

import { updatePersonDetail } from "../PersonDetail";

import styles from "./PeopleTable.module.scss";

/**
 * Star wars people table
 *
 * @returns Star wars people table in a Row
 */
function PeopleTable() {
  const { Column } = Table;
  const { Link, Title } = Typography;

  const [page, setPage] = useState(1);

  const { isLoading, isFetching, isSuccess, data } = useGetPeopleQuery(page);

  const dispatch = useAppDispatch();

  return (
    <Row justify="center">
      <Col className={styles.column_container}>
        <Skeleton
          active
          title={{ width: 200 }}
          paragraph={{ rows: 10, width: "100%" }}
          loading={isLoading}
          className={styles.skeleton}
        />

        {!isLoading && isSuccess && data && (
          <Table
            dataSource={data.results}
            rowKey="name"
            loading={isFetching}
            title={() => <Title level={4}>Star Wars People</Title>}
            bordered
            onRow={(record) => ({
              onClick: () =>
                dispatch(
                  updatePersonDetail({
                    name: record.name,
                    birthYear: record.birth_year,
                    gender: record.gender,
                    filmIDs: record.films.map((filmUrl) => {
                      if (typeof filmUrl === "string") {
                        return filmUrl
                          .replace("http://swapi.dev/api/films/", "")
                          .replace("/", "");
                      } else {
                        return "";
                      }
                    }),
                  })
                ),
            })}
            onChange={(pagination) => {
              dispatch(
                updatePersonDetail({
                  name: "",
                  birthYear: "",
                  gender: "",
                  filmIDs: [],
                })
              );
              setPage(pagination.current || 1);
            }}
            pagination={{
              current: page,
              total: data.count,
              showSizeChanger: false,
              itemRender: (current, type, originalElement) => {
                if (type === "prev") {
                  return <Link>Previous</Link>;
                }
                if (type === "next") {
                  return <Link>Next</Link>;
                }
                return originalElement;
              },
            }}
          >
            <Column title="Name" dataIndex="name" width={300} />
            <Column title="Height" dataIndex="height" width={300} />
            <Column title="Mass" dataIndex="mass" width={300} />
          </Table>
        )}
      </Col>
    </Row>
  );
}

export default PeopleTable;
