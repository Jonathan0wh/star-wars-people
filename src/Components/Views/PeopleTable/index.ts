export { default, updateCurrentList } from "./reducer";
