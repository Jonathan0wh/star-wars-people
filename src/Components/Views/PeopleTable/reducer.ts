import { createSlice, PayloadAction } from "@reduxjs/toolkit";

interface PeopleState {
  previousUrl: string | null;
  nextUrl: string | null;
  list: Array<{
    name: string;
    height: string;
    mass: string;
    birthYear: string;
    gender: string;
    filmsUrl: Array<string>;
  }>;
}

const initialState: PeopleState = {
  previousUrl: null,
  nextUrl: null,
  list: [],
};

export const peopleSlice = createSlice({
  name: "people",
  initialState,
  reducers: {
    updateCurrentList: (state, action: PayloadAction<PeopleState>) => {
      state = action.payload;
    },
  },
});

export const { updateCurrentList } = peopleSlice.actions;

export default peopleSlice.reducer;
