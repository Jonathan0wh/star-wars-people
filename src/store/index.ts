export { store, persistor } from "./store";

export type { RootState, AppDispatch } from "./store";
