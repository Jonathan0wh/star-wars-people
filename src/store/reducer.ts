import { combineReducers } from "redux";
import { swApi } from "services/swApi";

export const rootReducer = combineReducers({
  [swApi.reducerPath]: swApi.reducer,
});
