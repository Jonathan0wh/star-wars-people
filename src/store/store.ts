import { configureStore } from "@reduxjs/toolkit";
import { setupListeners } from "@reduxjs/toolkit/query";
import { personDetailReducer } from "Components/Views/PersonDetail";
import { swApi } from "services/swApi";

export const store = configureStore({
  reducer: {
    [swApi.reducerPath]: swApi.reducer,

    personDetail: personDetailReducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(swApi.middleware),
});

setupListeners(store.dispatch);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
