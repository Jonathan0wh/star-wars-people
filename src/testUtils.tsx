// test-utils.js
import React from "react";
import { applyMiddleware, createStore } from "redux";
import thunk from "redux-thunk";
import { Provider } from "react-redux";
import { render as rtlRender, RenderOptions } from "@testing-library/react";
// Import your own reducer
import { rootReducer } from "store/reducer";

type renderProps = {
  initialState?: object;
  store?: ReturnType<typeof createStore>;
} & RenderOptions;

function render(
  ui: React.ReactElement<any, string | React.JSXElementConstructor<any>>,
  {
    initialState,
    store = createStore(rootReducer, initialState, applyMiddleware(thunk)),
    ...renderOptions
  }: renderProps = {}
) {
  return rtlRender(ui, {
    wrapper: ({ children }: { children?: React.ReactNode }) => (
      <Provider store={store}>{children}</Provider>
    ),
    ...renderOptions,
  });
}

// re-export everything
export * from "@testing-library/react";
// override render method
export { render };
