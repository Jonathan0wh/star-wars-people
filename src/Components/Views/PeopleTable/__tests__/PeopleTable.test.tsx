import { store } from "store";
import { render } from "testUtils";

import PeopleTable from "../PeopleTable";

describe("View Component: PeopleTable", () => {
  const scrollToPersonDetail = jest.fn();

  it("should render by default", () => {
    const component = render(
      <PeopleTable scrollToPersonDetail={scrollToPersonDetail} />
    );

    const tree = component.asFragment();
    expect(tree).toMatchSnapshot();
  });

  it("should render with correct store data", () => {
    const component = render(
      <PeopleTable scrollToPersonDetail={scrollToPersonDetail} />,
      { store }
    );

    const tree = component.asFragment();
    expect(tree).toMatchSnapshot();
  });
});
