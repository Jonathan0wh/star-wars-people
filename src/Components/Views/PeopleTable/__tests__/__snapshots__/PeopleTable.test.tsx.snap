// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View Component: PeopleTable should render by default 1`] = `
<DocumentFragment>
  <div
    class="ant-row ant-row-center"
  >
    <div
      class="ant-col column_container"
    >
      <div
        class="ant-skeleton ant-skeleton-active skeleton"
      >
        <div
          class="ant-skeleton-content"
        >
          <h3
            class="ant-skeleton-title"
            style="width: 200px;"
          />
          <ul
            class="ant-skeleton-paragraph"
          >
            <li />
            <li />
            <li />
            <li />
            <li />
            <li />
            <li />
            <li />
            <li />
            <li
              style="width: 100%;"
            />
          </ul>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`View Component: PeopleTable should render with correct store data 1`] = `
<DocumentFragment>
  <div
    class="ant-row ant-row-center"
  >
    <div
      class="ant-col column_container"
    >
      <div
        class="ant-skeleton ant-skeleton-active skeleton"
      >
        <div
          class="ant-skeleton-content"
        >
          <h3
            class="ant-skeleton-title"
            style="width: 200px;"
          />
          <ul
            class="ant-skeleton-paragraph"
          >
            <li />
            <li />
            <li />
            <li />
            <li />
            <li />
            <li />
            <li />
            <li />
            <li
              style="width: 100%;"
            />
          </ul>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
